packages/cosmic-swingset/src/launch-chain.js

```diff
@@ -733,6 +742,66 @@ export async function launch({
         return true;
       }
 
+      case ActionType.ENACTED_UPGRADE: {
+        // Install and execute new core proposals.
+        const {
+          upgradePlan: { info: upgradeInfoJson = null } = {},
+          blockHeight,
+          blockTime,
+        } = action;
+        const upgradePlanInfo = upgradeInfoJson && JSON.parse(upgradeInfoJson);
+
+        // Handle the planned core proposals as just another action.
+        const { coreProposals = [] } = upgradePlanInfo || {};
+
+        if (!coreProposals.length) {
+          // Nothing to do.
+          return undefined;
+        }
+
+        const { bundles, code: coreEvalCode } =
+          await extractCoreProposalBundles(coreProposals, vatconfig, {
+            handleToBundleSpec: async (handle, source, _sequence, _piece) => {
+              const bundle = await bundleSource(source);
+              const { endoZipBase64Sha512: hash } = bundle;
+              const bundleID = `b1-${hash}`;
+              handle.bundleID = bundleID;
+              harden(handle);
+              return harden([handle.bundleID, bundle]);
+            },
+          });
+
+        for (const bundle of Object.values(bundles)) {
+          await controller.validateAndInstallBundle(bundle);
```

Comment:
It would be useful to thread some sort of name (that is not the bundle ID) for the bundle through `validateAndInstallBundle` in this case so that the errors say which bundle was bad. That would require a new argument and to thread that argument instead of the bundle ID into `checkBundle` when it is present and the other is absent.